testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_transactions.py

```diff
@@ -0,0 +1,420 @@
+
+    # 1: Insert a series of documents (A, B, C) in a transaction . CBL is not syncing as yet. SGW does import
+    transactions_app_dir = "resources/data/transaction-performer.jar"
+    doc_body = document.create_doc(doc_id=doc_id, content="doc1", channels=channels)
+    data = json.dumps(doc_body)
+    cmd = "java -cp {} com.couchbase.transaction clusterip={} operationdocids={} bucket={} doccontent='{}'".format(transactions_app_dir, c.servers[0].host, doc_ids, bucket, data)
+    subprocess.check_output(cmd, shell=True)
+
+    # 2. Start replication from SGW to CBL.
+    replicator = Replication(base_url)
+    sg_client.create_user(sg_admin_url, sg_db, username, password, channels=channels)
+    _, _, repl = replicator.create_session_configure_replicate(
+        base_url, sg_admin_url, sg_db, username, password, channels, sg_client, cbl_db, sg_blip_url, continuous=True, replication_type="push_pull")
+    replicator.wait_until_replicator_idle(repl)
+
+    # 3. Updates (edits ) a series of documents in this order (A->A',B-> B',C-> C') in a transaction . At the exact same time, CBL also edits the same document B -> B”
+    cbl_doc_ids = db.getDocIds(cbl_db)
+    cbl_db_docs = db.getDocuments(cbl_db, cbl_doc_ids)
+    doc_ids = doc_ids.replace("insert-", "replace-")
+    cmd = "java -cp {} com.couchbase.transaction clusterip={} operationdocids={} bucket={}".format(transactions_app_dir, c.servers[0].host, doc_ids, bucket)
+    with ThreadPoolExecutor(max_workers=2) as tpe:
+        shell_cmd_task = tpe.submit(shell_process, cmd)
+        for doc in cbl_db_docs:
+            if cbl_db_docs[doc]["id"] == cbl_update_id:
+                cbl_db_docs[doc]["content"] = "cbl_new_content"
+        db.updateDocuments(cbl_db, cbl_db_docs)
+        shell_cmd_task.result()
+
+    cbl_doc_ids = db.getDocIds(cbl_db)
+    cbl_docs = db.getDocuments(cbl_db, cbl_doc_ids)
+    for doc in cbl_docs:
+        if cbl_docs[doc]["id"] == cbl_update_id:
+            assert("_updated" in cbl_docs[doc]["id"])
+            assert("_updated" in cbl_docs[doc]["content"])
+
+    # 4. Do doc resurrection of one of the doc in transaction and verify resureccted doc replicated to cbl
+    resurrect_doc_ids = "remove-{}".format(resurrect_doc_id)
+    resurrect_insert_doc_ids = "{},insert-{}".format(resurrect_doc_ids, resurrect_doc_id)
+    cmd = "java -cp {} com.couchbase.transaction clusterip={} operationdocids={} bucket={}".format(transactions_app_dir, c.servers[0].host, resurrect_insert_doc_ids, bucket)
+    subprocess.check_output(cmd, shell=True)
+    cbl_doc_ids = db.getDocIds(cbl_db)
+    cbl_docs = db.getDocuments(cbl_db, cbl_doc_ids)
+    for doc in cbl_docs:
+        if cbl_docs[doc]["id"] == resurrect_doc_id:
+            assert("_updated" not in cbl_docs[doc]["id"])
+            assert("_updated" not in cbl_docs[doc]["content"])
+    replicator.stop(repl)
+
+
+def shell_process(cmd):
+    time.sleep(5)  # To make other process start before shell command executes
```

Comment:
popen helps here to avoid the sleep, since it initiates new process

Reply:
hmm... actually this sleep is to have cbl mutations to execute before cmd for transaction app is hit.  when I said other process, it is cbl mutations which is running in another thread as i need to run in two different process at the same time.  popen is just regular shell command , nothing to do with sleep requirement I added.